File: go/src/prefix/prefix128/prefix128.go
package prefix

// import "./prefix"
import "../ip_bits"
import "fmt"

///
///  Creates a new prefix object for 128 bits IPv6 addresses
///
///    prefix = IPAddressPrefix128.new 64
///      ///  64
///
func New(num uint8) (*Prefix, *string) {
    if num <= 128 {
        //static _FROM: &'static (Fn(&Prefix, usize)(*Prefix, *string)) = &from;
        //static _TO_IP_STR: &'static (Fn(&Vec<u16>)String) = &Prefix128::to_ip_str;
        ipBits := ip_bits.V6()
        bits := ipBits.Bits;
        return &Prefix {
            num,
            ipBits,
            prefix.New_netmask(num, bits),
            from, // vt_to_ip_str: _TO_IP_STR
        }, nil;
    }
    return nil, fmt.Sprintf("Prefix must be in range 0..128, got: %d", num);
}

func From(my Prefix, num uint) (*Prefix, *string) {
    return New(num);
}

File: kotlin/src/com/adviser/ipaddress/Prefix128.kt
package com.adviser.ipaddress



public class Prefix128 {

    companion object {

        public fun create(num: Int): Result<Prefix> {
            if (num <= 128) {
                //static _FROM: &'static (Fn(&Prefix, usize) -> Result<Prefix, String>) = &from;
                //static _TO_IP_STR: &'static (Fn(&Vec<u16>) -> String) = &Prefix128::to_ip_str;
                val ip_bits = IpBits.V6;
                val bits = ip_bits.bits;
                return Result.Ok(Prefix(
                        num,
                        ip_bits,
                        Prefix.new_netmask(num, bits),
                        { p: Prefix, _num: Int -> create(_num) }));
            }
            return Result.Err("Prefix must be in range 0..128, got: ${num}")
        }


        ///
        ///  Creates a new prefix object for 128 bits IPv6 addresses
        ///
        ///    prefix = IPAddressPrefix128.new 64
        ///      ///  64
        ///

        public fun from(num: Int): Result<Prefix> {
            return create(num);
        }
    }
}
